### PR TITLE
cache tls session in http1client

### PR DESCRIPTION
**Pull request by i110**
In http1client, cache tls session in h2o_cache_t and reuse it.

Related to:
https://github.com/h2o/h2o/issues/889
https://github.com/h2o/h2o/pull/875

### CODE REVIEW

File: lib/common/http1client.c [hunk 1]
```diff
@@ -380,6 +382,25 @@ static void on_connection_ready(struct st_h2o_http1client_private_t *client)
     h2o_timeout_link(client->super.ctx->loop, client->super.ctx->io_timeout, &client->_timeout);
 }
 
+SSL_SESSION *fetch_ssl_session_cache(h2o_http1client_t *client)
+{
+    h2o_cache_ref_t *ref = h2o_cache_fetch(client->ctx->ssl_session_cache, client->ctx->loop->_now, client->ssl.session_cache_key, client->ssl.session_cache_key_hash);
```
Comment (syohex) [inline lib/common/http1client.c]: `h2o_now` should be called instead of using `_now` field directly. `h2o_loop_t` when using `libuv` does not have `_now` field.

**Comment by kazuho** [comment]
Thank you for the PR. This is something we should have for H2O.

I would like to first discuss about the design before going into the details.

How about passing `h2o_cache_t *` as an argument to `h2o_socket_ssl_handshake`?

To me implementing session cache within the socket abstraction layer makes more sense, since I'd assume that not only http1client but other protocol implementations would want to use the cache.

**Comment by i110** [comment]
@kazuho Sure, I'll try it.

**Comment by i110** [comment]
@kazuho I addressed the points you said.

File: lib/common/socket.c [hunk 2]
```diff
@@ -828,6 +833,18 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
         }
     }
 
+    if (sock->ssl->handshake.client.session_cache) {
```
Comment (kazuho) [inline lib/common/socket.c]: I'd appreciate if the `if` statement was written to evaluate a boolean value, e.g. by comparing a pointer against `!= NULL`.

File: lib/common/http1client.c [hunk 3]
```diff
@@ -528,8 +528,9 @@ void h2o_http1client_connect(h2o_http1client_t **_client, void *data, h2o_http1c
 void h2o_http1client_connect_with_pool(h2o_http1client_t **_client, void *data, h2o_http1client_ctx_t *ctx,
                                        h2o_socketpool_t *sockpool, h2o_http1client_connect_cb cb)
 {
+
     struct st_h2o_http1client_private_t *client =
-        create_client(_client, data, ctx, sockpool->is_ssl ? sockpool->peer.host : h2o_iovec_init(NULL, 0), cb);
+        create_client(_client, data, ctx, sockpool->is_ssl ? sockpool->peer.host : h2o_iovec_init(NULL, 9), cb);
```
Comment (kazuho) [inline lib/common/http1client.c]: What does this `9` do?

File: lib/common/socket.c [hunk 4]
```diff
@@ -828,6 +833,18 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
...
+        if (err == NULL || err == h2o_socket_error_ssl_cert_name_mismatch) {
+            SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+            h2o_cache_set(sock->ssl->handshake.client.session_cache, h2o_now(h2o_socket_get_loop(sock)),
+                          sock->ssl->handshake.client.session_cache_key, sock->ssl->handshake.client.session_cache_key_hash,
+                          h2o_iovec_init(session, 1));
+        } else {
+            h2o_cache_delete(sock->ssl->handshake.client.session_cache, h2o_now(h2o_socket_get_loop(sock)),
+                             sock->ssl->handshake.client.session_cache_key, sock->ssl->handshake.client.session_cache_key_hash);
```
Comment (kazuho) [inline lib/common/socket.c]: Is there a need to remove a session cache upon failure?

The reason I ask this is there could be other reasons than the server loosing necessary information to resume a session that could cause the handshake to fail (e.g. TCP connection reset), in which case we _might_ not want to remove the cache entry.

File: lib/common/socket.c [hunk 5]
```diff
@@ -934,7 +951,21 @@ static void proceed_handshake(h2o_socket_t *sock, const char *err)
     on_handshake_complete(sock, err);
 }
 
-void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *server_name, h2o_socket_cb handshake_cb)
+h2o_iovec_t get_ssl_session_cache_key(h2o_socket_t *sock, const char *server_name)
```
Comment (kazuho) [inline lib/common/socket.c]: Maybe add `static`?

File: lib/common/socket.c [hunk 6]
```diff
@@ -934,7 +951,21 @@ static void proceed_handshake(h2o_socket_t *sock, const char *err)
...
+{
+    struct sockaddr sa;
+    h2o_socket_getpeername(sock, &sa);
+    uint16_t port = ntohs(((struct sockaddr_in *)&sa)->sin_port);
```
Comment (kazuho) [inline lib/common/socket.c]: There is no guarantee that `h2o_socket_getpeername` would succeed, nor that the returned type is `sockaddr_in`.

I believe you should for IPv4 and IPv6 addresses extract the port numbers (in appropriate ways), while refusing to use session resumption in all other cases.

And considering these requirements you _might_ want to reconsider how the generation of the key is extracted to a dedicated function.

File: lib/common/socket.c [hunk 7]
```diff
@@ -934,7 +951,21 @@ static void proceed_handshake(h2o_socket_t *sock, const char *err)
...
+
+    h2o_iovec_t key;
+    size_t len = strlen(server_name) + sizeof(H2O_UINT16_LONGEST_STR) + 1;
+    key.base = h2o_mem_alloc(len + 1);
+    key.len = snprintf(key.base, len + 1, "%s:%u", server_name, port);
```
Comment (kazuho) [inline lib/common/socket.c]: I assume your compiler reported an warning by using `%u` for `uint16_t`?

File: lib/common/socket.c [hunk 8]
```diff
@@ -960,6 +991,20 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
...
+            h2o_iovec_t session_cache_key = get_ssl_session_cache_key(sock, server_name);
+            sock->ssl->handshake.client.session_cache_key = session_cache_key;
+            sock->ssl->handshake.client.session_cache_key_hash = h2o_cache_calchash(session_cache_key.base, session_cache_key.len);
+            fprintf(stderr, "@@@ session_cache_key: %s\n", session_cache_key.base); // FIXME: nagata
+
+            h2o_cache_ref_t *cacheref =
+                h2o_cache_fetch(session_cache, h2o_now(h2o_socket_get_loop(sock)), sock->ssl->handshake.client.session_cache_key,
+                                sock->ssl->handshake.client.session_cache_key_hash);
+            if (cacheref != NULL) {
+                SSL_set_session(sock->ssl->ssl, (SSL_SESSION *)cacheref->value.base);
```
Comment (kazuho) [inline lib/common/socket.c]: A non-NULL reference returned by `h2o_cache_fetch` should be released by calling `h2o_cache_release`.

File: lib/common/socket.c [hunk 9]
```diff
@@ -999,6 +1044,17 @@ void h2o_socket_ssl_async_resumption_setup_ctx(SSL_CTX *ctx)
...
+static void on_session_cache_entry_destroy(h2o_iovec_t value)
+{
+    SSL_SESSION *session = (SSL_SESSION *)value.base;
+    SSL_SESSION_free(session);
+}
+
+h2o_cache_t *h2o_socket_ssl_create_session_cache(int flags, size_t capacity, uint64_t duration)
+{
+    return h2o_cache_create(flags, capacity, duration, on_session_cache_entry_destroy);
+}
```
Comment (kazuho) [inline lib/common/socket.c]: If we are to provide abstraction of the cache (i.e. this create function), then the type of the cache should be a type that wraps `h2o_cache_t`, instead of `h2o_cache_t` itself.

Instead, if we are to consider the type of the cache to be `h2o_cache_t`, the dispose callback should be exposed as a global function, and the user of the socket implementation should (be able to) construct the cache without using an abstraction.

For simplicity, my preference goes to the latter.

**Comment by kazuho** [comment]
Thank you for the changes. Aside from the points left in-line, looks good to me.

**Comment by i110** [comment]
@kazuho Thank you for your comments. I've addressed your comments, I believe.

File: lib/common/socket.c [hunk 10]
```diff
@@ -992,17 +976,27 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
...
-            fprintf(stderr, "@@@ session_cache_key: %s\n", session_cache_key.base); // FIXME: nagata
-
-            h2o_cache_ref_t *cacheref =
-                h2o_cache_fetch(session_cache, h2o_now(h2o_socket_get_loop(sock)), sock->ssl->handshake.client.session_cache_key,
-                                sock->ssl->handshake.client.session_cache_key_hash);
-            if (cacheref != NULL) {
-                SSL_set_session(sock->ssl->ssl, (SSL_SESSION *)cacheref->value.base);
+            struct sockaddr sa;
+            uint16_t port;
+            if (h2o_socket_getpeername(sock, &sa) != 0 && (port = h2o_socket_getport(&sa)) > 0) {
```
Comment (kazuho) [inline lib/common/socket.c]: You should check against `-1`, since the doc-comment of `h2o_socket_getport` suggests that the function _returns the port number, or -1 if failed_.

Note also that zero is a valid port number.

File: lib/common/socket.c [hunk 11]
```diff
@@ -992,17 +976,27 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
...
+                /* session cache is available */
+                h2o_iovec_t session_cache_key;
+                size_t len = strlen(server_name) + sizeof(H2O_UINT16_LONGEST_STR) + 1;
```
Comment (kazuho) [inline lib/common/socket.c]: Please change the calculation to `strlen(server_name) + sizeof(":" H2O_UINT16_LONGEST_STR)` following the representation of the string being built.

File: lib/common/socket.c [hunk 12]
```diff
@@ -992,17 +976,27 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
...
+                session_cache_key.base = h2o_mem_alloc(len + 1);
+                session_cache_key.len = snprintf(session_cache_key.base, len + 1, "%s:%" PRIu16, server_name, port);
```
Comment (kazuho) [inline lib/common/socket.c]: Let's just use `sprintf`, considering the fact that we are sure the generated string fits in memory.

File: lib/common/socket.c [hunk 13]
```diff
@@ -833,15 +833,12 @@ static void on_handshake_complete(h2o_socket_t *sock, const char *err)
         }
     }
 
-    if (sock->ssl->handshake.client.session_cache) {
-        if (err == NULL || err == h2o_socket_error_ssl_cert_name_mismatch) {
-            SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+    if (sock->ssl->handshake.client.session_cache != NULL) {
+        SSL_SESSION *session = SSL_get1_session(sock->ssl->ssl);
+        if (session != NULL) {
```
Comment (kazuho) [inline lib/common/socket.c]: How about _not_ calling `h2o_cache_set` unless the handshake succeeds?

The intent behind [my previous comment](https://github.com/h2o/h2o/pull/1053#discussion_r76928596) was that the session cache should be updated (either by calling _set_ or _delete_) only after a successful connection establishment.

File: include/h2o/socket.h [hunk 14]
```diff
@@ -278,9 +279,11 @@ int32_t h2o_socket_getport(struct sockaddr *sa);
  * performs SSL handshake on a socket
  * @param sock the socket
  * @param ssl_ctx SSL context
+ * @param session_cache SSL session cache (maybe NULL)
```
Comment (kazuho) [inline include/h2o/socket.h]: I would appreciate it if the type of the entries within the cache (i.e. `SSL_SESSION*`) would be clarified here.

**Comment by kazuho** [comment]
Thank you for the changes. I've left my comments in-line, please let me know what you think.

**Comment by i110** [comment]
@kazuho Thank you for the review. I fixed the problems.
BTW, I wonder if I should set a cache entry when `h2o_socket_error_ssl_cert_name_mismatch` error occured and the verify mode is `SSL_VERIFY_NONE`. WDYT?

File: lib/handler/proxy.c [hunk 15]
```diff
@@ -96,6 +96,7 @@ static void on_context_init(h2o_handler_t *_self, h2o_context_t *ctx)
         client_ctx->websocket_timeout = NULL;
     }
     client_ctx->ssl_ctx = self->config.ssl_ctx;
+    client_ctx->ssl_session_cache = h2o_cache_create(0, 4096, 86400 * 1000, h2o_socket_ssl_destroy_session_cache_entry);
```
Comment (kazuho) [inline lib/handler/proxy.c]: Don't we need to set `H2O_CACHE_FLAG_MULTITHREADED` here as well?

**Comment by kazuho** [comment]
@i110 
> BTW, I wonder if I should set a cache entry when `h2o_socket_error_ssl_cert_name_mismatch` error occured and the verify mode is `SSL_VERIFY_NONE`. WDYT?

I believe we should always update the cache when the result is `h2o_socket_error_ssl_cert_name_mismatch` (i.e. even if the verify mode is not `SSL_VERIFY_NONE`), considering the fact that it is not actually an error caused _below_ the TLS handshake layer.

**Comment by kazuho** [comment]
LGTM aside from my comment in-line.

@i110 Before merging this, I would like to see configuration directives introduced for the session cache (e.g. size (could be zero to disable), lifetime).

And regarding the fact, it might be better to discuss how the directives should look like before you start implementing it. Could you please make a proposal?

File: lib/handler/proxy.c (hunk 15 above)
Comment (i110) [inline lib/handler/proxy.c]: In my understanding,  `on_context_init` is called for each thread, so this session cache never be accessed from other threads. Is this wrong?

Comment (kazuho) [inline lib/handler/proxy.c]: Oh, I missed the fact that the function is thread-specific.

So let me rephrase the question to: why are we creating a cache per every thread? I believe it should be shared between the worker threads.

**Comment by i110** [comment]
@kazuho 
> And regarding the fact, it might be better to discuss how the directives should look like before you start implementing it. Could you please make a proposal?

I guess it could be like the following:

```
# identical to the default value when the directive doesn't exist
proxy.ssl.session-cache:
  capacity: 4096
  duration: 24 hour # same format as the "expires" directive

# simply disable the feature
proxy.ssl.session-cache: OFF
```

**Comment by kazuho** [comment]
@i110 Thank you for the proposal.

It might be enough to just allow numbers as the argument to `duration` in the unit of seconds, considering the fact that we have similar patterns everywhere.

**Comment by i110** [comment]
@kazuho https://github.com/h2o/h2o/pull/1053/commits/e1aec8af29a3510d903d15f269f7e497c0af3acf adds the configuration directive. I changed `duration` to `lifetime` to use same vocabulary as `ssl-session-resumption`

File: lib/handler/configurator/proxy.c [hunk 16]
```diff
@@ -142,6 +142,52 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+    struct proxy_configurator_t *self = (void *)cmd->configurator;
+    size_t i;
+
+    switch (node->type) {
+    case YOML_TYPE_SCALAR:
+        if (strcasecmp(node->data.scalar, "OFF") == 0) {
+            /* disabled */
+            self->vars->session_cache.capacity = 0;
+        } else if (strcasecmp(node->data.scalar, "ON") == 0) {
+            /* use default values */
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Does this work if for an example the feature is turned off at the global level and then set to `ON` at the host level?

File: lib/handler/configurator/proxy.c [hunk 17]
```diff
@@ -142,6 +142,52 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+            yoml_t *value = node->data.mapping.elements[i].value;
+            if (key->type == YOML_TYPE_SCALAR) {
+                if (strcasecmp(key->data.scalar, "capacity") == 0) {
+                    if (h2o_configurator_scanf(cmd, value, "%" PRIu64, &self->vars->session_cache.capacity) != 0)
+                        return -1;
+                } else if (strcasecmp(key->data.scalar, "lifetime") == 0) {
+                    if (h2o_configurator_scanf(cmd, value, "%" PRIu32, &self->vars->session_cache.lifetime) != 0)
+                        return -1;
+                } else {
+                    goto InvalidKey;
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Please please do not abuse goto. Goto should never be used for jumping into another block.

**Comment by kazuho** [comment]
Thank you for the updates. Aside from my two comments in-line please consider https://github.com/h2o/h2o/pull/1053#discussion_r77115971.

File: lib/handler/configurator/proxy.c (hunk 17 above)
Comment (i110) [inline lib/handler/configurator/proxy.c]: ><

**Comment by kazuho** [comment]
PS. Also please write tests that confirm the use of session resumption (or no use, depending on the configuration).

There'd be several ways to write such test. For example, you can run another H2O server as upstream, and collect the value of `%{ssl.session-reused}x` from the access log.

File: lib/handler/configurator/proxy.c (hunk 16 above)
Comment (i110) [inline lib/handler/configurator/proxy.c]: I forgot the fact that `self->vars` inherits parent values.. I'll fix.

File: lib/handler/proxy.c (hunk 15 above)
Comment (i110) [inline lib/handler/proxy.c]: I guess I misunderstood the meaning of what you said: `h2o_http1client_ctx_t ごとに h2o_cache_t を作るという実装でとりあえずいいんじゃないか`. I'll fix.

File: lib/handler/configurator/proxy.c [hunk 18]
```diff
@@ -142,6 +149,71 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
     return ret;
 }
 
+static h2o_cache_t *create_ssl_session_cache(uint64_t capacity, unsigned lifetime)
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Type of `capacity` should match that of `h2o_cache_create` (and generally speaking, capacity is bound to by the maximum size of memory that can be allocated, thus should be `size_t`).

File: lib/handler/configurator/proxy.c [hunk 19]
```diff
@@ -142,6 +149,71 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+{
+    return h2o_cache_create(H2O_CACHE_FLAG_MULTITHREADED, capacity, lifetime * 1000, h2o_socket_ssl_destroy_session_cache_entry);
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: `lifetime * 1000` should be calculated using `uint64_t` (which is the type accepted by `h2o_cache_create). Otherwise, you are risking a integer overflow (that leads to undefined behavior).

File: lib/handler/configurator/proxy.c [hunk 20]
```diff
@@ -142,6 +149,71 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+    case YOML_TYPE_MAPPING:
+        for (i = 0; i != node->data.mapping.size; ++i) {
+            yoml_t *key = node->data.mapping.elements[i].key;
+            yoml_t *value = node->data.mapping.elements[i].value;
+            if (key->type == YOML_TYPE_SCALAR) {
+                if (strcasecmp(key->data.scalar, "capacity") == 0) {
+                    if (h2o_configurator_scanf(cmd, value, "%" PRIu64, &capacity) != 0)
+                        return -1;
+                } else if (strcasecmp(key->data.scalar, "lifetime") == 0) {
+                    if (h2o_configurator_scanf(cmd, value, "%" PRIu32, &lifetime) != 0)
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Type specifier for `unsigned` is not `PRIu32`.

File: lib/common/socket.c [hunk 21]
```diff
@@ -999,6 +1039,37 @@ void h2o_socket_ssl_async_resumption_setup_ctx(SSL_CTX *ctx)
...
+static void on_dispose_ssl_ctx_ex_data(void *parent, void *ptr, CRYPTO_EX_DATA *ad, int idx, long argl, void *argp)
+{
+    h2o_cache_t *ssl_session_cache = (h2o_cache_t *)ptr;
+    if (ssl_session_cache != NULL)
+        h2o_cache_destroy(ssl_session_cache);
+}
+
+static int get_ssl_session_cache_index()
+{
+    static int index = INT_MAX;
```
Comment (kazuho) [inline lib/common/socket.c]: Initial value should be set to a value that is never returned by `SSL_CTX_get_ex_new_index`. Please consult the manual of the function to determine the appropriate value.

File: lib/common/socket.c [hunk 22]
```diff
@@ -999,6 +1039,37 @@ void h2o_socket_ssl_async_resumption_setup_ctx(SSL_CTX *ctx)
...
+    if (index == INT_MAX)
+        index = SSL_CTX_get_ex_new_index(0, NULL, NULL, NULL, on_dispose_ssl_ctx_ex_data);
```
Comment (kazuho) [inline lib/common/socket.c]: Please check the returned value (the check could be an `assert`).

File: lib/common/socket.c (hunk 21 above)
Comment (kazuho) [inline lib/common/socket.c]: Also, please make this function multi-thread safe. I'd suggest using double-checked locking for the matter.

File: lib/handler/configurator/proxy.c [hunk 23]
```diff
@@ -99,6 +100,10 @@ static void update_ssl_ctx(SSL_CTX **ctx, X509_STORE *cert_store, int verify_mod
     CRYPTO_add(&cert_store->references, 1, CRYPTO_LOCK_X509_STORE);
     if (verify_mode == -1)
         verify_mode = (*ctx)->verify_mode;
+    if (session_cache == NULL && delete_session_cache == 0) {
+        session_cache = h2o_socket_ssl_get_session_cache(*ctx);
+        h2o_socket_ssl_set_session_cache(*ctx, NULL); /* prevent the cache from being disposed */
+    }
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Sorry, I do not get this.

My understanding is that it is a requirement to have a session cache associated for every SSL_CTX that has been set up with both capacity and lifetime set to a non-zero value.

However, in this _if_ block, you seem to be detaching an session cache from an existing SSL_CTX.

Doesn't this cause a problem?

As an example, consider the following configuration. Would upstream connections for a.example.com use the session cache?

```yaml
proxy.ssl-session-cache: ON
hosts:
  a.example.com:
    /:
      proxy.reverse.url: ...
  b.example.com:
    /:
      proxy.reverse.url: ...
    proxy.ssl.session-cache: OFF
```

**Comment by kazuho** [comment]
@i110 Thank you for the updates. I've left a few comments. Please let me know what you think.

File: lib/handler/configurator/proxy.c (hunk 19 above)
Comment (i110) [inline lib/handler/configurator/proxy.c]: honto da.

File: lib/handler/configurator/proxy.c (hunk 23 above)
Comment (i110) [inline lib/handler/configurator/proxy.c]: this is absolutely my mistake.. sorry..

**Comment by i110** [comment]
@kazuho Thank you for the review. I addressed the issues, and add tests.

**Comment by kazuho** [comment]
Thank you for the updates. Can you please address https://github.com/h2o/h2o/pull/1053#discussion_r77501192 as well?

File: lib/common/socket.c [hunk 24]
```diff
@@ -999,6 +1039,39 @@ void h2o_socket_ssl_async_resumption_setup_ctx(SSL_CTX *ctx)
...
 }
 
+static void on_dispose_ssl_ctx_ex_data(void *parent, void *ptr, CRYPTO_EX_DATA *ad, int idx, long argl, void *argp)
+{
+    h2o_cache_t *ssl_session_cache = (h2o_cache_t *)ptr;
+    if (ssl_session_cache != NULL)
+        h2o_cache_destroy(ssl_session_cache);
+}
+
+static int get_ssl_session_cache_index()
```
Comment (kazuho) [inline lib/common/socket.c]: In C, you should always specify `(void)` when declaring a function that takes zero arguments (notice that this line works as both declaration and definition).

You may interested in reading http://d.hatena.ne.jp/eel3/20141005/1412521223 to understand the reason behind the necessity.

File: lib/handler/configurator/proxy.c [hunk 25]
```diff
@@ -99,6 +100,13 @@ static void update_ssl_ctx(SSL_CTX **ctx, X509_STORE *cert_store, int verify_mod
     CRYPTO_add(&cert_store->references, 1, CRYPTO_LOCK_X509_STORE);
     if (verify_mode == -1)
         verify_mode = (*ctx)->verify_mode;
+    if (session_cache == NULL && delete_session_cache == 0) {
+        session_cache = h2o_socket_ssl_get_session_cache(*ctx);
+    }
+    if ((*ctx)->references == 1) {
+        /* ctx will be disposed, but prevent the cache from being disposed by setting NULL to reuse it later*/
+        h2o_socket_ssl_set_session_cache(*ctx, NULL);
+    }
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: I don't think this is the right change.

It seems to me that with this change, the function might associate one session cache to more than one SSL_CTX, if `(*ctx)->references` is greater than one.

As discussed in my previous comment, the rule we agreed is that the relation between `SSL_CTX` vs a session cache should be either 1:0 or 1:1.

IMO the easiest way to fulfill such rule would be to create and (optionally) associate the session cache at the time when and only when SSL_CTX is created within this function.

File: lib/handler/configurator/proxy.c [hunk 26]
```diff
@@ -142,6 +152,77 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+
+static int on_config_ssl_session_cache(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node)
+{
+    struct proxy_configurator_t *self = (void *)cmd->configurator;
+    size_t i;
+    size_t capacity = self->vars->ssl_session_cache.capacity;
+    unsigned lifetime = self->vars->ssl_session_cache.lifetime;
+    if (capacity == 0 || lifetime == 0) {
+        capacity = H2O_DEFAULT_PROXY_SSL_SESSION_CACHE_CAPACITY;
+        lifetime = H2O_DEFAULT_PROXY_SSL_SESSION_CACHE_LIFETIME;
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Throughout the PR, either `capacity` or `lifetime` being zero is used as an indication of the cache being disabled.

The approach seems unnecessary complex to me.

How about using specifying either of the value being zero as a mark for indicating a disabled cache?

Also, it might also be a good idea to emit an error in case either of the two is set to zero but the other is not to avoid confusion, instead of silently disabling the cache.

File: t/50reverse-proxy-session-resumption.t [hunk 27]
```diff
@@ -130,13 +130,16 @@ EOC
...
+    my $app_server = spawn_h2o(<< "EOT");
+num-threads: 2
+hosts:
+  default:
+    paths:
+      /:
+        header.add: "X-Reproxy-URL: https://127.0.0.1:$upstream_port"
+        file.dir: @{[ DOC_ROOT ]}
+EOT
+    my $app_port = $app_server->{port};
```
Comment (kazuho) [inline t/50reverse-proxy-session-resumption.t]: Or maybe use `plackup -s Starlet`? My understanding is that we do that in other places, and generally _not_ relying on H2O act as a tool to provide test data would be a good idea (considering the fact that the intention of the test is to develop H2O).

File: lib/common/socket.c [hunk 28]
```diff
@@ -977,9 +977,9 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
     } else {
         h2o_cache_t *session_cache = h2o_socket_ssl_get_session_cache(ssl_ctx);
         if (session_cache != NULL) {
-            struct sockaddr sa;
+            struct sockaddr_storage sa;
             int32_t port;
-            if (h2o_socket_getpeername(sock, &sa) != 0 && (port = h2o_socket_getport(&sa)) != -1) {
+            if (h2o_socket_getpeername(sock, (struct sockaddr *)&sa) != 0 && (port = h2o_socket_getport((struct sockaddr *)&sa)) != -1) {
```
Comment (kazuho) [inline lib/common/socket.c]: Off topic, but we might consider changing the signature of `h2o_socket_getpeername` to take three arguments instead of two, following the API of `getsockname`.

File: include/h2o.h [hunk 29]
```diff
@@ -370,6 +372,10 @@ struct st_h2o_globalconf_t {
          */
         SSL_CTX *ssl_ctx;
         /**
+         * SSL session cache (optional, governed by the application)
+         */
+        h2o_cache_t *ssl_session_cache;
```
Comment (kazuho) [inline include/h2o.h]: Is this value ever used?

Comment (i110) [inline include/h2o.h]: NO!

File: lib/handler/configurator/proxy.c [hunk 30]
```diff
@@ -89,14 +89,13 @@ static SSL_CTX *create_ssl_ctx(void)
     return ctx;
 }
 
-static h2o_cache_t *create_ssl_session_cache(size_t capacity, unsigned lifetime)
+static h2o_cache_t *null_session_cache = (h2o_cache_t *){0}; /* indicate to make session cache empty */
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: I am very opposed to this. It is so easy to misuse a special value like `null_session_cache`.

As an alternative approach, how about passing `h2o_cache_t **session_cache` as an argument to `update_ssl_ctx`? If `session_cache` is NULL, then the function inherits the current parameters. Otherwise, session cache of the updated SSL_CTX is set to the value of `*session_cache`.

Comment (i110) [inline lib/handler/configurator/proxy.c]: sonna wazaga..

File: lib/handler/configurator/proxy.c [hunk 31]
```diff
@@ -99,6 +104,13 @@ static void update_ssl_ctx(SSL_CTX **ctx, X509_STORE *cert_store, int verify_mod
     CRYPTO_add(&cert_store->references, 1, CRYPTO_LOCK_X509_STORE);
     if (verify_mode == -1)
         verify_mode = (*ctx)->verify_mode;
+    h2o_cache_t *new_session_cache;
+    if (session_cache == NULL) {
+        h2o_cache_t *current = h2o_socket_ssl_get_session_cache(*ctx);
+        new_session_cache = create_ssl_session_cache(h2o_cache_get_capacity(current), h2o_cache_get_duration(current));
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Isn't there a case where `current` is NULL?

Comment (i110) [inline lib/handler/configurator/proxy.c]: ossharu toori.

**Comment by i110** [comment]
@kazuho Thank you for your checks. I fixed the issue, and also I merged master branch and resolved some conflicts.

File: lib/handler/configurator/proxy.c [hunk 32]
```diff
@@ -140,6 +156,85 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
...
+    struct proxy_configurator_t *self = (void *)cmd->configurator;
+    size_t i;
+    size_t capacity = 0;
+    uint64_t duration = 0;
+    h2o_cache_t *current_cache = h2o_socket_ssl_get_session_cache(self->vars->ssl_ctx);
+
+    switch (node->type) {
+    case YOML_TYPE_SCALAR:
+        if (strcasecmp(node->data.scalar, "OFF") == 0) {
+            if (current_cache) {
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Use `!= NULL` to make the result a boolean?

File: lib/handler/configurator/proxy.c [hunk 33]
```diff
@@ -140,6 +156,85 @@ static int on_config_ssl_cafile(h2o_configurator_command_t *cmd, h2o_configurato
     return ret;
 }
 
+static int on_config_ssl_session_cache(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node)
+{
+    struct proxy_configurator_t *self = (void *)cmd->configurator;
+    size_t i;
```
Comment (kazuho) [inline lib/handler/configurator/proxy.c]: Move the declaration to inside of `case YOML_TYPE_MAPPING: {`?

File: t/50reverse-proxy-session-resumption.t [hunk 34]
```diff
@@ -0,0 +1,211 @@
...
+    unless prog_exists('curl');
+plan skip_all => 'Starlet not found'
+    unless system('perl -MStarlet /dev/null > /dev/null 2>&1') == 0;
+
+my $tempdir = tempdir(CLEANUP => 1);
+
+sub create_upstream {
+    unlink "$tempdir/access_log";
+    return spawn_h2o(<< "EOT");
+num-threads: 2
```
Comment (kazuho) [inline t/50reverse-proxy-session-resumption.t]: Could you please explain why you need to set this?

**Comment by kazuho** [comment]
Thank you for the changes. LGTM aside from my comments in-line.

File: t/50reverse-proxy-session-resumption.t (hunk 34 above)
Comment (i110) [inline t/50reverse-proxy-session-resumption.t]: there's no need, so I'll remove it

File: lib/common/socket.c [hunk 35]
```diff
@@ -963,6 +978,32 @@ void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *
...
+        h2o_cache_t *session_cache = h2o_socket_ssl_get_session_cache(ssl_ctx);
+        if (session_cache != NULL) {
+            struct sockaddr_storage sa;
+            int32_t port;
+            if (h2o_socket_getpeername(sock, (struct sockaddr *)&sa) != 0 &&
+                (port = h2o_socket_getport((struct sockaddr *)&sa)) != -1) {
+                /* session cache is available */
+                h2o_iovec_t session_cache_key;
+                size_t len = strlen(server_name) + sizeof(":" H2O_UINT16_LONGEST_STR);
+                session_cache_key.base = h2o_mem_alloc(len + 1);
```
Comment (i110) [inline lib/common/socket.c]: Sure, you're right and I agree with your idea.

**Comment by i110** [comment]
@kazuho Thank you for pointing out the issues. I fixed.

**Comment by kazuho** [comment]
Thank you for the fixes! Merged to master.
